File: 01 - Data Preprocessing/data_preprocessing_template.R
# Data Preprocessing Template

# Importing the dataset
dataset = read.csv('Data.csv')

# Taking care of missing values
dataset$Age[is.na(dataset$Age)] <- mean(dataset$Age, na.rm=TRUE)
dataset$Salary[is.na(dataset$Salary)] <- mean(dataset$Salary, na.rm=TRUE)

# Encode categorical data
dataset$Country = factor(dataset$Country, # i dati da codificare
                         levels = c('France', 'Spain', 'Germany'), # c(..) specifica un vettore
                         labels = c(1, 2, 3)) # label assegnati a France, Spain, Germany
dataset$Purchased = factor(dataset$Purchased,
                           levels = c('Yes', 'No'),
                           labels = c(1, 0))
# in R dopo l'encoding non è vero che 1, 2, 3 hanno pesi diversi

# Splitting the dataset into the Training set and Test set
install.packages('caTools')
library(caTools)
set.seed(123) # inizializza il generatore di numeri random
split = sample.split(dataset$Purchased, SplitRatio = 0.8) # si scrive la funzione split per poterla usare dopo
                                                          # bisogna passare come parametro solo le var DIPENDENTI
                                                          # splitratio indica la percentuale dedicata al training set
training_set = subset(dataset, split == TRUE)     # il training set si compone dei valori agli indici di 'split' che valgono True (stampa 'split' nel terminale per capire che sto dicendo)
test_set = subset(dataset, split == FALSE)        # il test set si compone dei valori agli indici false dell'array 'split'

# Feature Scaling
training_set[, 2:3] = scale(training_set[, 2:3]) # applicachiamo il rescaling solo dove è necessario
test_set[, 2:3] = scale(test_set[, 2:3])
# se si verifica questo errore:
#   Error in colMeans(x, na.rm = TRUE) : 'x' must be numeric
# è perchè abbiamo inserito dummy value in colonne che contenevano del testo, e questi numeri sono rimasti testo.

File: 01 - Data Preprocessing/data_preprocessing_template.py
# Data Preprocessing Template

# Importing the libraries
import numpy as np # contiene le funzioni matematiche
import matplotlib.pyplot as plt # permette di fare i grafici
import pandas as pd # per importare e maneggiare dataset

# Importing the dataset
dataset = pd.read_csv('Data.csv')

# Categorical data
country_dummies = pd.get_dummies(dataset.Country, prefix="Country", drop_first=True)
purchased_dummies = pd.get_dummies(dataset.Purchased, prefix="Purchased", drop_first=True)
dataset = pd.concat([country_dummies, dataset], axis = 1)
dataset = pd.DataFrame.drop(dataset, columns=["Country"])
dataset = pd.concat([dataset, purchased_dummies], axis = 1)
dataset = pd.DataFrame.drop(dataset, columns=["Purchased"])

# Variabili dipendenti e indipendenti
X = dataset.iloc[:, :-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()

# Taking care of missing data
from sklearn.impute import SimpleImputer
simple_imputer = SimpleImputer(missing_values=np.nan, strategy="mean")
simple_imputer.fit(X[:, 2:4]) # simple_imputer viene fatto lavorare solo nelle colonne dove ci sono dati mancanti
X[:, 2:4] = simple_imputer.transform(X[:, 2:4])

# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
stdscaler_X = StandardScaler()
X_train = stdscaler_X.fit_transform(X_train)
X_test = stdscaler_X.transform(X_test)
# sc_y = StandardScaler()
# y_train = sc_y.fit_transform(y_train)